### DISCUSSION TITLE
Error logs

### DISCUSSION THREAD

**Discussion by s-canchi** [nih-cfde/training-and-engagement#436] [Snakemake] [answered]
Is there a way to capture the errors from the command(s) in the Snakemake rule to output somewhere else or change the message?

**Comment by marisalim** [comment, marked as answer]
One way to capture the standard error output is to add a `log` directive in your snakemake rules. For example, in the `map_reads` rule we wrote in the Snakemake workshop, we can add a `log` block and modify the `bwa mem` command to save log files in a `logs/bwa_mem/` directory:

```
rule map_reads:
    input:
        genome="ecoli-rel606.fa",
        reads="SRR2584857_1.fastq.gz",
        pac="ecoli-rel606.fa.pac",
        ann="ecoli-rel606.fa.ann",
        amb="ecoli-rel606.fa.amb",
        bwt="ecoli-rel606.fa.bwt",
        sa="ecoli-rel606.fa.sa"
    output: sam="SRR2584857.sam"
    log: "logs/bwa_mem/ecoli-rel606.log"
    shell:
        "(bwa mem -t 4 {input.genome} {input.reads} > {output.sam}) 2> {log}"
```

The following is saved in the `ecoli-rel606.log` file if it completes successfully:
```
[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 100000 sequences (10100000 bp)...
[M::mem_process_seqs] Processed 100000 reads in 4.574 CPU sec, 1.172 real sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa mem -t 4 ecoli-rel606.fa SRR2584857_1.fastq.gz
[main] Real time: 1.576 sec; CPU: 4.885 sec
```

A failed output might look like this (here, the index input files were missing from the rule):
```
[E::bwa_idx_load_from_disk] fail to locate the index files
```

See the [snakemake documentation](https://snakemake.readthedocs.io/en/stable/tutorial/advanced.html#step-5-logging) for another example!
